### ISSUE TITLE
Rich UI and graphics with notcurses

### ISSUE THREAD

**Issue by olanod** [nushell/nushell#6222]
### Describe the solution you'd like

> [Notcurses](https://github.com/dankamongmen/notcurses) is a library facilitating complex TUIs on modern terminal emulators, supporting vivid colors, multimedia, threads, and Unicode to the maximum degree possible.

It would be interesting to use notcurses to support rich graphics, i.e. plotting data or `open` multimedia. Overtime the library could be used in more areas of nushell UI like in menus, prompts, etc.

**Comment by rgwood** [comment]
Notcurses is, indeed, extremely cool. I was looking into writing .NET bindings for it a while back (unrelated to Nushell). I'd recommend that anyone not familiar with it check out this demo video: https://www.youtube.com/watch?v=dcjkezf1ARY

I think we might be open to PRs for this, but we'd need answers to a few questions first:

- how hard is it to statically link Notcurses into a Rust app like Nushell?
- what kind of compile time and binary size penalty would we pay for embedding Notcurses?

It's also possible that Notcurses might make more sense in Nushell plugins than in Nushell's core.

**Comment by fdncred** [comment]
We used to have some plugins that used tui-rs I believe. I always fixed we'd return to tui-rs.

**Comment by olanod** [comment]
> I think we might be open to PRs for this, but we'd need answers to a few questions first:
> 
>     * how hard is it to statically link Notcurses into a Rust app like Nushell (and does it work on all 3 major OSs)?
> 
>     * what kind of compile time and binary size penalty would we pay for embedding Notcurses?
> 

I haven't tried it yet but there is a [`notcurses`](https://crates.io/crates/notcurses) crate that has most features and re-exports [`libnotcurses-sys`](https://crates.io/crates/libnotcurses-sys) to access whatever is not implemented yet. The creator of notcurses also seems [interested in Rust](https://youtu.be/mOmMcFXdd6k?t=1154) so I take that as a positive sign that support will continue to improve. So if static linking is not already the default for the sys crate it should be simple to add support for it.  
About the size I don't know what would be acceptable, what is mentioned in the readme is that,
> Notcurses can be built without its multimedia functionality, requiring a significantly lesser set of dependencies.
So if opening something like videos is not a priority then some bytes can be spared.

**Comment by fdncred** [comment]
LOL, the author in that video made me laugh by saying, "The rust community is a little huggy feely for my taste". 

> If PRs are welcomed...

That's a big yes, please submit a PR or 2 or 10 with this stuff. 😆 

> what kind of initial integration would be nice to see? perhaps rendering nushell on a notcurses plane ...

yes, that sounds interesting.

I was also thinking about something hopefully simple like `open cat.png` and it displays the image in the terminal for real kind of like with sixel or other terminal rendering technologies. I can do it in iterm2 with a script but it would be cool if it was native.

> reedline + nushell

I wonder if we could render some cool completion combo boxes or some cool UI listbox? It would be neat to see an experiment.

**Comment by rgwood** [comment]
> I was also thinking about something hopefully simple like open cat.png and it displays the image in the terminal for real kind of like with sixel or other terminal rendering technologies.

Yeah, I think a command that displays an image would be a good first experiment. That seems like one of Notcurses's strong suits, and it should (🤞) involve way fewer architectural changes than overhauling Nushell/Reedline's rendering.

**Comment by fdncred** [comment]
cool. is there a repo where i can look at the code. i've been interested in drawing pics for a while. i've successfully been able to use the kitty and iterm2 protocols to draw pics but i'm guessing this may be something independent of that?

**Comment by fdncred** [comment]
very interesting. It would be cool to have a nu_plugin_<some_imaging_name> that displays images as an experimental nushell command.

**Comment by fdncred** [comment]
oh, wow. Those look pretty pixelated but cool!

**Comment by Fl1tzi** [comment]
Yeah it's because I'm using Alacritty. In Kitty you have the "native quality" image viewing.

**Comment by fdncred** [comment]
ya, we'll want to test a variety of terminals on Windows, MacOS, and Linux eventually. That's why i was thinking that a experimental plugin could probably be ideal for this type of thing so we can get a minimum viable candidate in and the people an test it and contribute to it if they're interested.

**Comment by Fl1tzi** [comment]
Notcurses does provide image viewing for Iterm2 so mac would also get the great viewing. Every other terminal that does not provide that type of support for pixels is able to just run the most compatible blitter.

**Comment by tryggvil** [comment]
Viu (https://github.com/atanunq/viu) or more specifically its underlying library (https://github.com/atanunq/viuer) could also be a more minimal option for displaying images. It supports the Kitty/iTerm2 image protocol as well as optionally Sixel

**Comment by olanod** [comment]
Viu looks nice, I'd say if we only end up using notcurses for pictures(and video?) then sure, a lighter weight library is better. But the point of notcurses is to eventually use more of its capabilities to draw more parts of the interface.
